File: modules/wem-admin-ui/src/main/resources/web/admin/common/js/content/page/region/ComponentType.ts
module api.content.page.region {

    export class ComponentType {

        private static shortNameToInstance: {[shortName: string]: ComponentType} = {};

        private shortName: string;

        constructor(shortName: string) {
            ComponentType.shortNameToInstance[shortName] = this;
            this.shortName = shortName;
        }

        getShortName(): string {
            return this.shortName;
        }

        newComponentBuilder(): ComponentBuilder<Component> {
            throw new Error("Must be implemented by inheritors");
        }

        static byShortName(shortName: string): ComponentType {
            return ComponentType.shortNameToInstance[shortName];
        }
    }

}

File: modules/core/core-api/src/main/java/com/enonic/xp/region/ComponentType.java
package com.enonic.xp.region;

public abstract class ComponentType
{
    private final String shortName;

    public ComponentType( final String shortName )
    {
        this.shortName = shortName;
    }

    @Override
    public String toString()
    {
        return shortName;
    }
}
